op-chain-ops/cmd/celo-migrate/beacon.go

```diff
@@ -15,71 +16,166 @@ import (
+// block is found it will consider future epochs that may have finalized a block
+// ocurring before the given time.
+func (c *BeaconClient) MostRecentFinalizedBlockAtTime(chainID *big.Int, unixTime uint64) (common.Hash, error) {
+	var l1StartBlockHash common.Hash
+	var l1StartBlockTime uint64
+	// This loop looks back for a finalized L1 block that is up to maxSequencerDrift before the L2 fork block.
+	for epoch := ContainingEpoch(chainID, unixTime); ; epoch-- {
+		AwaitEpoch(chainID, epoch)
+		slot := FirstSlotOfEpoch(epoch)
+		finalityCheckpoints, err := c.FindFinalityCheckpointsForSlot(slot, 10)
+		if err != nil {
+			if errors.Is(err, ethereum.NotFound) {
+				continue
+			}
+			return common.Hash{}, fmt.Errorf("failed fetching finality checkpoints for slot %d: %w", slot, err)
+		}
+		justifiedEpoch := uint64(finalityCheckpoints.Data.Finalized.Epoch)
+		if !withinMaxSequencerDrift(EpochStartTime(chainID, justifiedEpoch), unixTime) {
+			// We've gone too far back and not found a suitable block, so break.
+			break
+		}
+
+		finalizedSlot := FirstSlotOfEpoch(justifiedEpoch)
+		l1StartBlockHash, l1StartBlockTime, err = c.FindBlockForSlot(finalizedSlot, 10)
+		if err != nil {
+			if errors.Is(err, ethereum.NotFound) {
+				continue
+			}
+			return common.Hash{}, fmt.Errorf("failed fetching L1 block for slot (%v): %w", finalizedSlot, err)
+		}
+		if !withinMaxSequencerDrift(l1StartBlockTime, unixTime) {
+			// This block is too old to use with the L2 fork block. Nullify the hash
+			// to signify that no suitable block was yet found.
+			l1StartBlockHash = common.Hash{}
+		} else {
+			log.Info(fmt.Sprintf("Found finalized L1 slot at %d", finalizedSlot))
+		}
+		break
+	}
+
+	// If no block found start looking to future epochs, failing if we
+	// cannot find a finalized block ocurring before the L2 fork block.
+	if l1StartBlockHash == (common.Hash{}) {
+		for epoch := ContainingEpoch(chainID, unixTime) + 1; ; epoch++ {
+			AwaitEpoch(chainID, epoch)
+			slot := FirstSlotOfEpoch(epoch)
+			finalityCheckpoints, err := c.FindFinalityCheckpointsForSlot(slot, 10)
+			if err != nil {
+				if errors.Is(err, ethereum.NotFound) {
+					continue
+				}
+				return common.Hash{}, fmt.Errorf("failed fetching finality checkpoints for slot %d: %w", slot, err)
+			}
+			justifiedEpoch := uint64(finalityCheckpoints.Data.Finalized.Epoch)
+			finalizedSlot := FirstSlotOfEpoch(justifiedEpoch)
+			l1StartBlockHash, l1StartBlockTime, err = c.FindBlockForSlot(finalizedSlot, 10)
+			if err != nil {
+				if errors.Is(err, ethereum.NotFound) {
+					continue
+				}
+				return common.Hash{}, fmt.Errorf("failed fetching L1 block for slot (%v): %w", finalizedSlot, err)
+			}
+			if l1StartBlockTime >= unixTime {
+				// We've gone too far forward and not found a suitable block, so nullify the hash and break.
+				l1StartBlockHash = common.Hash{}
+			} else {
+				log.Info(fmt.Sprintf("Found finalized L1 slot at %d", finalizedSlot))
+			}
+			break
+		}
```

Comment:
This is very similar to above, maybe it makes sense to put this in a function that has a parameter where you can pass in a function to check the validity of the blocktime.